src/profile/context/__tests__/ProfileContext.test.tsx

```diff
@@ -0,0 +1,95 @@
+
+describe('ProfileContext', () => {
+  function createTestEnv(responses: MockedResponse[]) {
+    const responseProvider: ResponseProvider = () =>
+      responses.shift() as MockedResponse;
+    return exposeProfileContext(responseProvider);
+  }
+
+  afterEach(() => {
+    cleanup();
+    cleanupHooks();
+    resetApolloMocks();
+  });
+
+  it('should have no data before fetch is called and all trackers are "false" ', async () => {
+    const { result } = createTestEnv([]);
+    const context = result.current;
+    expect(context.data).toBeUndefined();
+    expect(context.loading).toEqual(false);
+    expect(context.isInitialized).toEqual(false);
+    expect(context.isComplete).toEqual(false);
+  });
+
+  it("after fetch(), context indicates 'loading' state and data updates when fetch is finished", async () => {
+    const responses: MockedResponse[] = [
+      { profileData: getMyProfile().myProfile as ProfileData },
+    ];
+    const { result, waitForUpdate } = createTestEnv(responses);
+    let context = result.current;
+    await act(async () => {
+      const loadingPromise = waitForUpdate();
+      context.fetch();
+      await loadingPromise;
+      context = result.current;
+      expect(context.data).toBeUndefined();
+      expect(context.loading).toEqual(true);
+      expect(context.isInitialized).toEqual(true);
+      expect(context.isComplete).toEqual(false);
+
+      const dataLoadedPromise = waitForUpdate();
+      await dataLoadedPromise;
+      context = result.current;
+      expect(context.data).toBeDefined();
+      expect(context.loading).toEqual(false);
+      expect(context.isInitialized).toEqual(true);
+      expect(context.isComplete).toEqual(true);
+      expect(context.data?.myProfile?.firstName).toEqual('Teemu');
+      expect(context.getName()).toEqual('Teemu Testaaja');
+      expect(context.getName(true)).toEqual('Teme');
+    });
+  });
+  it('Fetch errors are handled and listeners triggered and disposed', async () => {
+    const responses: MockedResponse[] = [
+      { errorType: 'graphQLError' },
+      { errorType: 'networkError' },
+    ];
+    const { result, waitForErrorChange } = createTestEnv(responses);
+    let context = result.current;
+    const errorListener = jest.fn();
+    await act(async () => {
+      // first promise is resolved when context.loading changes to true
+      const errorPromise = waitForErrorChange();
+      const listenerDisposer = context.addErrorListener(errorListener);
+      context.fetch();
+      await errorPromise;
+      context = result.current;
+      expect(context.data).toBeUndefined();
+      expect(context.error).toBeDefined();
+      expect(context.loading).toEqual(false);
+      expect(context.isInitialized).toEqual(true);
+      expect(context.isComplete).toEqual(false);
+      // apollo triggers error listeners later, "waitForTimeout" added 'cause of that
+      await waitForTimeout(100);
+      expect(errorListener.mock.calls.length).toEqual(1);
```

Comment:
Suggesting to replace a fixed delay with waiting for a condition:

```suggestion
      await waitFor(() => expect(errorListener.mock.calls.length).toEqual(1));
```

 Needs importing `waitFor` from `@testing-library/react`.

Reply:
Changed